### ISSUE TITLE
new version nnUNet?

### ISSUE THREAD

**Issue by alizezhu** [MIC-DKFZ/nnUNet#184]
Hi Fabian,

Thanks for your fabulous nnUNet! I found that you have recently updated it. I have been working with the previous version for some time. Is there a need to update to the new version?

Also I have tried to work with the new version, and I found that in the new version I got warnings that I did not have before, including

Warning:  multi_tensor_applier fused unscale kernel is unavailable, possibly because apex was installed without --cuda_ext --cpp_ext. Using Python fallback.  Original ImportError was: ModuleNotFoundError("No module named 'amp_C'")

and during training:

Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 32768.0

Is that a big problem and would it impact the performance of segmentation?

Thanks!

**Comment by alizezhu** [comment]
Hi Fabian,
Another question: a bit counter-intuitively, I found that running time of an epoch is not related to the training data size: for example, Task02_Heart took longer than Task01_BrainTumour per epoch. Could you give me a hint? :-)

**Comment by alizezhu** [comment]
I tried the same dataset on two versions of nnUNet:

old

epoch:  0
2020-04-04 13:23:49.936361: train loss : -0.3579
2020-04-04 13:24:16.108433: val loss (train=False): -0.6760
2020-04-04 13:24:16.141833: Val glob dc per class: [0.8794355870742593]
2020-04-04 13:24:18.756796: lr is now (scheduler) 0.0003
2020-04-04 13:24:18.760370: current best_val_eval_criterion_MA is 0.87940
2020-04-04 13:24:18.762733: current val_eval_criterion_MA is 0.8794
2020-04-04 13:24:18.764833: No improvement: current train MA -0.3579, best: -0.3579, eps is 0.0005
2020-04-04 13:24:18.766889: Patience: 0/50
2020-04-04 13:24:18.768916: This epoch took 581.416025 s



new nnUNet


2020-04-04 13:30:04.933073:
epoch:  0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 32768.0
2020-04-04 13:41:29.345903: train loss : -0.5676
2020-04-04 13:42:06.809027: validation loss: -0.7777
2020-04-04 13:42:06.824502: Average global foreground Dice: [0.8406887713348726]
2020-04-04 13:42:06.837406: (interpret this as an estimate for the Dice of the different classes. This is not exact.)
2020-04-04 13:42:15.606503: lr: 0.009991
2020-04-04 13:42:15.634966: This epoch took 721.888238 s



It seems the new version took longer on one epoch. I used the same dataset Task02_Heart. I wonder if there is some change in the training or is there something wrong with my implementation... Thanks again!

**Comment by FabianIsensee** [comment]
Hi,
the warning are not a problem. They come from apex/amp. 

> Warning: multi_tensor_applier fused unscale kernel is unavailable, possibly because apex was installed without --cuda_ext --cpp_ext. Using Python fallback. Original ImportError was: ModuleNotFoundError("No module named 'amp_C'")

This means that you did the python only install of apex. That is OK.

> Gradient overflow. Skipping step, loss scaler 0 reducing loss scale to 32768.0

This comes from mixed precision training, a new feature of nnUNet V2. You don't have to worry about that

Interesting to see that the new nnunet is a lot slower for you. I have not seen that in my experiments. Can you please run it with `--fp32` and report back how long an epoch takes with that?

On some datasets the new nnunet is substantially better than the old one. If you prefer to use the old one you can still do this after the update. Just specify `nnUNetTrainer` instead of nnUNetTrainerV2. You will also need to do a different preprocessing. Let me know if you really want to do that and I will create a guide for that.

Best,
Fabian

**Comment by alizezhu** [comment]
Fabian, thanks! I tried to add --fp32 in the argument, but it seems that the speed is the same.

Previously I found that the time for training an epoch does not seem to correlate with the training data size: for example, for Task01_BrainTumor which has much larger data size than Task02_Heart, it trains actually faster. Also if I train two versions fullres and lowres on the same dataset, the speed is the same. Is that correct also in your experience?

For the versions, if there is a short intro on where the differences are in implementation (like data preparation and command line) that would be very helpful. Thanks!

**Comment by alizezhu** [comment]
GPU is this
![image](https://user-images.githubusercontent.com/8353480/78460899-2d5e2180-76c5-11ea-946d-078e70331744.png)

**Comment by FabianIsensee** [comment]
Hi,
the time for each epoch should be roughly the same for all datasets (except hippocampus) due to the way nnU-Net handles 'epochs' and GPU loading (always max out the gpu budget).

I find it curious that nnunetV2 takes longer per epoch on your system. Could it be that you are running into a CPU or I/O bottleneck? Can you please check the CPU load during training? I am also using Titan Xp GPUs and my epoch times are a shorter.

Regarding the deployment: It's so slow because it is using sliding window prediction and test time data augmentation. It was not designed for speed because I prioritized segmentation accuracy. If you want to sped it up, the first thing to try would be to disable test time data augmentation (spedup of roughtly factor 8). 

Best,
Fabian

**Comment by alizezhu** [comment]
Hi Fabian, Thanks for your reply! I wonder if I can train multiple folds in parallel on different cores to save time. I guess the splitting between the training and validation is fixed.

**Comment by FabianIsensee** [comment]
You mean train at the same time on different GPUs? Yes this should work now

**Comment by alizezhu** [comment]
Fabian, I tested the new version again on a number of datasets, the new version on my core is consistently a bit slower than the previous version, around  10% - 20% more time. (e.g. previously 320 s, now 380 s) I guess this is due to the warning? :(

May I ask how long it takes per epoch on your V100?

Warning: multi_tensor_applier fused unscale kernel is unavailable, possibly because apex was installed without --cuda_ext --cpp_ext. Using Python fallback. Original ImportError was: ModuleNotFoundError("No module named 'amp_C'")

**Comment by FabianIsensee** [comment]
Hi,
what hardware are you running this on?
Best,
Fabian

**Comment by alizezhu** [comment]
NVIDIA Tesla V100, 16GB memory.

**Comment by FabianIsensee** [comment]
Then it should be about twice as fast as before. Please verify whether you are having a CPU bottleneck

**Comment by alizezhu** [comment]
Hi Fabian, thanks. Can you let me know how to check the CPU bottleneck? I am working on a Centos server

**Comment by FabianIsensee** [comment]
You sh into it and run htop while the training is running :-)
That's how I do it.
What is the hardware configuration of the server? What CPU is it using (lscpu) and how many GPUs are installed?

**Comment by shifuxiao** [comment]
Hi Fabian, thanks,I use my dataset ,but dice list none value.Can you help me work on it.
![image](https://user-images.githubusercontent.com/65807507/188090783-a1b40186-3c4a-4183-b231-9b702ab16776.png)

**Comment by shifuxiao** [comment]
Hi Fabian, thanks,I use my dataset ,but dice list none value.Can you help me work on it.
![image](https://user-images.githubusercontent.com/65807507/188092649-1e669afd-2c2e-4ab4-9d15-b300dcd2db80.png)
